Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalPipeline.cpp

```diff
@@ -0,0 +1,4371 @@
+        fDevice.CurrentRenderCommandEncoder()->setVertexBuffer(vRef->GetBuffer(), 0, 0);
+        fState.fCurrentVertexBuffer = vRef->GetBuffer();
+    }
+
+    // Only render projections and shadows if we successfully rendered the span.
+    // j == -1 means we aborted render.
+    if (pass >= 0) {
+        // if we had to render aux spans, we probably changed the vertex and index buffer
+        // reset those
+        fState.fCurrentVertexBuffer = vRef->GetBuffer();
+        fDevice.fCurrentIndexBuffer = iRef->GetBuffer();
+
+        // Projections that get applied to the frame buffer (after all passes).
+        if (fProjAll.size() && !(fView.fRenderState & kRenderNoProjection)) {
+            fDevice.CurrentRenderCommandEncoder()->pushDebugGroup(MTLSTR("Render All Projections"));
+            IRenderProjections(render, vRef);
+            fDevice.CurrentRenderCommandEncoder()->popDebugGroup();
+        }
+
+        // Handle render of shadows onto geometry.
+        if (fShadows.size()) {
+            IRenderShadowsOntoSpan(render, &span, material, vRef);
+        }
+    }
+
+    if (span.GetNumAuxSpans() || (pass >= 0 && fShadows.size())) {
+    }
+
+#ifdef HS_DEBUGGING
+    fDevice.CurrentRenderCommandEncoder()->popDebugGroup();
+#endif
+}
+
+// IRenderProjections ///////////////////////////////////////////////////////////
+// Render any projected lights that want to be rendered a single time after
+// all passes on the object are complete.
+void plMetalPipeline::IRenderProjections(const plRenderPrimFunc& render, const plMetalVertexBufferRef* vRef)
+{
+    PushCurrentLightSources();
+    IDisableLightsForShadow();
+    for (plLightInfo* li : fProjAll) {
+        IRenderProjection(render, li, vRef);
+    }
+    PopCurrentLightSources();
+}
+
+// IRenderProjection //////////////////////////////////////////////////////////////
+// Render this light's projection onto the frame buffer.
+void plMetalPipeline::IRenderProjection(const plRenderPrimFunc& render, plLightInfo* li, const plMetalVertexBufferRef* vRef)
+{
+    // Enable the projecting light only.
+    IEnableLight(0, li);
+    fLights.count = 1;
+
+    plLayerInterface* proj = li->GetProjection();
+    CheckTextureRef(proj);
+    plMetalTextureRef* tex = (plMetalTextureRef*)proj->GetTexture()->GetDeviceRef();
+
+    IScaleLight(0, true);
+
+    fCurrentRenderPassUniforms->ambientSrc = 1.0;
+    fCurrentRenderPassUniforms->diffuseSrc = 1.0;
+    fCurrentRenderPassUniforms->emissiveSrc = 1.0;
+    fCurrentRenderPassUniforms->specularSrc = 1.0;
+    fCurrentRenderPassUniforms->globalAmb = {1.0, 1.0, 1.0};
+    fCurrentRenderPassUniforms->ambientCol = {0.0, 0.0, 0.0};
+    fCurrentRenderPassUniforms->emissiveCol = {0.0, 0.0, 0.0};
+    fCurrentRenderPassUniforms->specularCol = {0.0, 0.0, 0.0};
+    fCurrentRenderPassUniforms->fogColor = {0.0, 0.0, 0.0};
+    fCurrentRenderPassUniforms->diffuseCol = {1.0, 1.0, 1.0, 1.0};
```

Comment:
The types here don't seem to match up? `ambientSrc` is a integer type, not a floating point. The colors don't seem to use doubles, either.

Reply:
The colors use halves - which I don't think I can have literals for in C++. But I'll reduce them to floats.